### PR TITLE
Logging

### PR DESCRIPTION
**Pull request by janosdebugs**
## Please describe the change you are making

This PR adds logging to the image upload process via the [go-ovirt-client-log library](https://github.com/ovirt/go-ovirt-client-log). It is based on #5 .

### CODE REVIEW

File: client_disk_uploadimage.go [hunk 1]
```diff
@@ -135,21 +136,31 @@ func (o *oVirtClient) createProgress(
...
 }
 
 type uploadImageProgress struct {
-	cli             *oVirtClient
-	uploadedBytes   uint64
-	cowSize         uint64
-	size            uint64
-	reader          *bufio.Reader
+	// cli is the ovirtClient that initiated the upload
+	cli *oVirtClient
```
Comment (Gal-Zaidman) [inline client_disk_uploadimage.go:147]: Not related to loging

File: client_disk_uploadimage.go [hunk 2]
```diff
@@ -327,41 +345,43 @@ func (u *uploadImageProgress) findTransferURL(transfer *ovirtsdk4.ImageTransfer)
...
-		transferURL, err := url.Parse(transferURL)
+		parsedTransferURL, err := url.Parse(transferURL)
 		if err != nil {
 			lastError = wrap(err, EUnidentified, "failed to parse transfer URL %s", transferURL)
 			continue
 		}
 
-		hostUrl, err := url.Parse(transfer.MustTransferUrl())
+		// TODO retry options request a few times in case of a network failure.
+		optionsReq, err := http.NewRequest(http.MethodOptions, parsedTransferURL.String(), strings.NewReader(""))
```
Comment (Gal-Zaidman) [inline client_disk_uploadimage.go:362]: Seems more like  a bug fix then logging

File: client_disk_uploadimage.go [hunk 3]
```diff
@@ -374,47 +394,53 @@ func (u *uploadImageProgress) findTransferURL(transfer *ovirtsdk4.ImageTransfer)
...
+	u.logger.Debugf("Creating disk for image upload...")
+	// TODO add retries for disk creation
 	addDiskRequest := u.conn.SystemService().DisksService().Add().Disk(u.disk)
 	addDiskRequest.Query("correlation_id", u.correlationID)
 	addResp, err := addDiskRequest.Send()
 	if err != nil {
 		diskAlias, _ := u.disk.Alias()
-		return "", nil, wrap(err, EUnidentified, "failed to create disk, alias: %s", diskAlias)
+		err = wrap(err, EUnidentified, "failed to create disk, alias: %s", diskAlias)
+		u.logger.Debugf(fmt.Sprintf("%s", err))
```
Comment (Gal-Zaidman) [inline client_disk_uploadimage.go:405]: shouldn't that be err.string()
https://github.com/oVirt/go-ovirt-client/blob/main/errors.go#L97-L98

File: client_disk_uploadimage.go [hunk 4]
```diff
@@ -374,47 +394,53 @@ func (u *uploadImageProgress) findTransferURL(transfer *ovirtsdk4.ImageTransfer)
...
+		return transfer, transferService, err
+	}
+
+	if err := u.waitForImageTransferReady(diskID, transferService); err != nil {
+		return transfer, transferService, err
 	}
-	transfer = transferRes.MustImageTransfer()
-	transferService := imageTransfersService.ImageTransferService(transfer.MustId())
+	return transfer, transferService, nil
+}
```
Comment (Gal-Zaidman) [inline client_disk_uploadimage.go]: Not related to logging

File: client_disk_uploadimage.go [hunk 5]
```diff
@@ -327,41 +345,43 @@ func (u *uploadImageProgress) findTransferURL(transfer *ovirtsdk4.ImageTransfer)
...
+					lastError = newError(
+						EConnection,
+						"non-200 status code returned from URL %s (%d)",
+						parsedTransferURL.String(),
+						res.StatusCode,
+					)
 				}
-			} else {
-				lastError = err
 			}
```
Comment (Gal-Zaidman) [inline client_disk_uploadimage.go]: Seems more like a bug fix then logging

File: client_disk_uploadimage.go [hunk 6]
```diff
@@ -424,30 +450,93 @@ func (u *uploadImageProgress) setupImageTransfer(diskID string) (
...
+
+	transfer, ok := transferRes.ImageTransfer()
+	if !ok {
+		return nil, nil, newError(EFieldMissing, "missing image transfer as a response to image transfer create request")
+	}
+	transferID, ok := transfer.Id()
+	if !ok {
+		return nil, nil, newError(EFieldMissing, "missing image transfer ID in response to image transfer create request")
+	}
+	transferService := imageTransfersService.ImageTransferService(transferID)
```
Comment (Gal-Zaidman) [inline client_disk_uploadimage.go]: Not related to logging

**Review by Gal-Zaidman — CHANGES_REQUESTED** [review]
PR looks good, I think that we will probably add more logs as we will refactor the code.
But there is a large part of this PR that is not related to logging, its more of a refactor/bugfix change.
I think we should split this into 2 PRs or at least 2 commits (and not squash merge)

**Comment by janosdebugs** [comment]
Thanks @Gal-Zaidman the issues should now be resolved.

**Comment by Gal-Zaidman** [comment]
> Thanks @Gal-Zaidman the issues should now be resolved.

I still see that everything is the same

**Review by Gal-Zaidman — APPROVED** [review]

**Pull request merged**